[3.6 only] Document transitional feature macros / guards

mpg opened the pull request:
## Description

Resolves #8736

## PR checklist

- [x] **changelog** not required because: documentation
- [x] **development PR** not required because: 4.0 will complete the transition and have much simpler guards
- [x] **framework PR** not required
- [x] **3.6 PR** here
- [x] **2.28 PR** not required because: 2.28 does not have those guards that were added for driver-only support
- **tests**  not required: documentation

gilles-peskine-arm commented on docs/architecture/psa-migration/strategy.md:
```diff
@@ -18,11 +18,15 @@ needs to be changed to use new APIs. For a more detailed account of what's
 
-As of early 2023, work towards G5 is in progress: Mbed TLS 3.3 and 3.4 saw
-some improvements in this area, and more will be coming in future releases.
+As of Mbed TLS 3.6 (early 2024, work towards G5 is well advanced: it is now
```
```suggestion
As of Mbed TLS 3.6 (early 2024), work towards G5 is well advanced: it is now
```

gilles-peskine-arm commented on docs/architecture/psa-migration/strategy.md:
```diff
@@ -18,11 +18,15 @@ needs to be changed to use new APIs. For a more detailed account of what's
-some improvements in this area, and more will be coming in future releases.
+As of Mbed TLS 3.6 (early 2024, work towards G5 is well advanced: it is now
+possible to have hashes/HMAC, ciphers/AEAD, and ECC provided only by drivers,
+with some limitations. See `docs/driver-only-builds.md` for details.
```
```suggestion
with some limitations. See [`docs/driver-only-builds.md`](../../driver-only-builds.html) for details.
```

gilles-peskine-arm commented on docs/architecture/psa-migration/strategy.md:
```diff
@@ -219,7 +223,12 @@ Strategies currently (early 2022) used with each abstraction layer:
+- Cipher (G1, TLS): replace calls at each call site
+- Cipher (G5): create a new internal abstraction layer for (non-DES) block
+  ciphers that silently calls PSA when a driver is available, see
+  `md-cipher-dispatch.md`.
```
```suggestion
  [`md-cipher-dispatch.md`](md-cipher-dispatch.html).
```

gilles-peskine-arm reviewed (changes requested):
Looks correct to me apart from some minor niggles.

However, this does not address https://github.com/Mbed-TLS/mbedtls/issues/8736.

mpg commented:
> However, this does not address #8736.

Aw, I forgot to add the new file, how embarrassing. Sorry.

gilles-peskine-arm commented on docs/architecture/psa-migration/transition-guards.md:
```diff
@@ -0,0 +1,290 @@
+- transitional `MBEDTLS_xxx` macros that stem from the desire to be able to
+  use crypto mechanisms that are only provided by a driver (G5 in
+`strategy.md`).
+
```
Macros defined in `include/mbedtls/config_adjust_legacy_*.h` only, used outside of `include/mbedtls/*.h` and not mentioned here:

* `MBEDTLS_BLOCK_CIPHER_xxx_VIA_xxx` (used in library, test dependencies)
* `MBEDTLS_BLOCK_CIPHER_SOME_PSA` (used in library, test helpers, test code)
* `MBEDTLS_MD_xxx_VIA_PSA` (used in library, test dependencies)
* `MBEDTLS_PSA_UTIL_HAVE_ECDSA` (used in library, test dependencies)

We should document them as well, but that's not a blocker for this pull request.

gilles-peskine-arm commented:
This should be on the 3.6 branch, since it won't be relevant in 4.0.

mpg commented:
@gilles-peskine-arm I added commits addressing your feedback, and then rebased to resolve conflicts and more importantly target the correct branch. Please feel free to review again. Note: I wasn't sure if this is priority medium of high, please feel free to adjust.

mpg commented:
> @mpg 1. So this shouldn't be a draft PR anymore? 2. The CI isn't happy.

Yeah, I meant to un-draft it but forgot, thanks for the reminder. Regarding the CI, I'm not sure what happened but it's looking happy now.

gilles-peskine-arm commented on docs/architecture/psa-migration/psa-limitations.md:
```diff
@@ -14,8 +14,8 @@ Limitations relevant for G1 (performing crypto operations)
 
-- Restartable/interruptible ECC operations: in progress (mid-2024).
-- Arbitrary parameters for FFDH: will be dropped in 4.0.
+- Restartable/interruptible ECC operations: support in PSA will be added in 4.0.
```
“will be added in 4.0” isn't quite accurate. Interruptible signatures were already in 2.28. Restartable ECDH should be in 4.0. We haven't decided yet about restartable export-public. Their usage in TLS will not be in 4.0.

gilles-peskine-arm commented on docs/architecture/psa-migration/transition-guards.md:
```diff
@@ -0,0 +1,290 @@
+Ciphers (AEAD and unauthenticated)
+----------------------------------
+
+**Overview of existing (internal) APIs:** we currently have 4 (families of)
```
Block cipher modes are still missing herre.

gilles-peskine-arm commented on docs/architecture/psa-migration/transition-guards.md:
```diff
@@ -0,0 +1,295 @@
+domain, needs a hash. (Note that this includes `ENTROPY_C`, so in practice
+`MD_LIGHT` is enabled in most builds.)
+
+Note that there is a rule, enforced by `config_adjust_psa_superset_legacy.h`,
```
Not a blocker, but the clarifications I requested on this paragraph (https://github.com/Mbed-TLS/mbedtls/pull/9161/files#r1657124669, https://github.com/Mbed-TLS/mbedtls/pull/9161/files#r1657126157, https://github.com/Mbed-TLS/mbedtls/pull/9161/files#r1657136618https://github.com/Mbed-TLS/mbedtls/pull/9161/files#r1657136618) haven't been addressed.

gilles-peskine-arm commented on docs/architecture/psa-migration/transition-guards.md:
```diff
@@ -0,0 +1,290 @@
+are `GCM` and `CCM` which use the internal abstraction layer `block_cipher`
+and check for availability of block ciphers using `MBEDTLS_CCM_GCM_CAN_xxx`
+macros defined in `config_adjut_legacy_crypto.h`.
+
```
CTR_DRBG is still missing here.

gilles-peskine-arm commented on docs/architecture/psa-migration/transition-guards.md:
```diff
@@ -0,0 +1,290 @@
+Key derivation
+--------------
+
+**Legacy and `USE_PSA` domains:** no users here.
```
The legacy API does expose key derivation used by code in the USE_PSA domain: PKCS5 and PKCS12.

mpg replied:
Well, the key derivation functions are exposed but not directly called by library code - PK only uses those key derivations indirectly through the use of the associated password-based encryption scheme. I've added a paragraph anyway.

gilles-peskine-arm commented on docs/architecture/psa-migration/transition-guards.md:
```diff
@@ -0,0 +1,290 @@
+
+**Curves:** in `config_adjut_psa_superset_legacy.h` we ensure that all
+curves that are supported on the legacy side (`MBEDTLS_ECP_DP_xxx_ENABLED`)
+are also supported on the PSA side (`PSA_WANT_ECC_xxx`).
```
As noted before, only if some PSA is included.

gilles-peskine-arm reviewed (changes requested):
Thanks for the updates! This mostly looks good to me, but there's one 4.0 plan that just changed, and a few requests for clarifications from my previous review that you haven't addressed and I think at least some of them should be addressed.

mpg commented:
@gilles-peskine-arm Thanks for your review and sorry for forgetting to address some of your comments. I believe I've now addressed all of them except for https://github.com/Mbed-TLS/mbedtls/pull/9161#discussion_r1662777690 which I'd rather leave for later.

gilles-peskine-arm commented on docs/architecture/psa-migration/transition-guards.md:
```diff
@@ -88,22 +88,22 @@ domain, needs a hash. (Note that this includes `ENTROPY_C`, so in practice
 
 **Pure TLS 1.3 domain:** it is not easy to know which uses of hashes fall in
-this domain as opposed to the `USE_PSA` domain which looking at the code.
+this domain as opposed to the `USE_PSA` domain whithout looking at the code.
```
Trivial:
```suggestion
this domain as opposed to the `USE_PSA` domain without looking at the code.
```

gilles-peskine-arm reviewed (approved):
LGTM

Harry-Ramsey reviewed (approved):
LGTM. I don't have prior experience with 3.6 drivers but this information seems vastly useful for developing ontop of Mbed TLS. The transition guards section will be useful for developing ontop of Mbed TLS or contributing to the library as a whole.
